File: core/src/com/lastwords/states/GameStateManager.java
package com.lastwords.states;

import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.lastwords.ashley.velocity.VelocitySystem;

import java.util.Stack;

public class GameStateManager {

    private Stack<State> states;
    private Engine engine;
    private SpriteBatch spriteBatch;

    public GameStateManager(SpriteBatch spriteBatch) {
        this.engine = new Engine();
        this.states = new Stack<State>();
        this.spriteBatch = spriteBatch;
    }

    public void setMainSystems() {
        engine.addSystem(new VelocitySystem());
    }

    public void addEntity(Entity entity) {
        engine.addEntity(entity);
    }

    public void push(State state) {
        states.push(state);
    }

    public void pop() {
        states.pop().dispose();
    }

    public void set(State state) {
        pop();
        states.push(state);
    }

    public void update(float dt) {
        states.peek().update(dt);
    }

    public void renrer(SpriteBatch spriteBatch) {
        states.peek().render(spriteBatch);
    }

    public Engine getEngine() {
        return engine;
    }

    public SpriteBatch getSpriteBatch() {
        return spriteBatch;
    }

    public void setSpriteBatch(SpriteBatch spriteBatch) {
        this.spriteBatch = spriteBatch;
    }

    public void resize(int width, int height) {
        states.peek().resize(width, height);
    }
}

File: core/src/com/lastwords/states/GameStateManager.kt
package com.lastwords.states

import com.badlogic.ashley.core.Engine
import com.badlogic.ashley.core.Entity
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.lastwords.ashley.velocity.VelocitySystem

import java.util.Stack

class GameStateManager(var spriteBatch: SpriteBatch) {

    val states: Stack<State> = Stack()
    val engine: Engine = Engine()

    fun setMainSystems() {
        engine.addSystem(VelocitySystem())
    }

    fun addEntity(entity: Entity) {
        engine.addEntity(entity)
    }

    fun push(state: State) {
        states.push(state)
    }

    fun pop() {
        states.pop().dispose()
    }

    fun set(state: State) {
        pop()
        states.push(state)
    }

    fun update(dt: Float) {
        states.peek().update(dt)
    }

    fun render(spriteBatch: SpriteBatch) {
        states.peek().render(spriteBatch)
    }

    fun resize(width: Int, height: Int) {
        states.peek().resize(width, height)
    }
}
